### PR TITLE
Allowing for user defined transforms

### PR DESCRIPTION
**Pull request by edkazcarlson**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements and Flexibility in Data Handling and Model Configuration

### 📊 Key Changes
- Introduced `inputCh` configuration to specify the number of input channels for images.
- Added flexibility in data augmentation and label transformation with `override_label_transforms` and `append_label_transforms` options.
- Enhanced model and dataset initialization to support the new configurations.
- Improved plot handling for datasets with non-standard (non-3) image channels.

### 🎯 Purpose & Impact
- **Custom Input Channels**: Allows models to handle images with different numbers of channels (e.g., grayscale or multispectral images), making the library more versatile.
- **Data Augmentation Customization**: The addition of label transformation options provides users with the ability to customize or extend the data preprocessing and augmentation steps. This can lead to better model performance by tailoring the preprocessing steps to specific dataset characteristics.
- **Better Support for Non-Standard Images**: The updates ensure that plotting functions gracefully handle images with non-standard channels, avoiding errors and improving user experience when working with such data.

These changes enhance the library's flexibility, making it more adaptable to various types of data and specific project requirements.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added support for grayscale and custom channel images in YOLO model training and improved transformation handling.

### 📊 Key Changes
- **New Classes for Grayscale Conversion**: Added classes to handle grayscale conversion during classification, detection, and segmentation.
- **Additional Configuration Parameters**: Introduced `input_Ch` parameter to specify the number of input channels for non-standard images.
- **Enhanced Data Loading & Transformation**: Modified dataset building and data loading functions to support custom transformations and handle new channel configurations.
- **Updated Documentation**: Adjusted documentation to include new parameters and functionalities.

### 🎯 Purpose & Impact
- **Broader Image Support**: Allows training on grayscale and other non-RGB images by specifying the number of input channels.
- **Custom Transformations**: Enables users to append or override default transformations, providing greater flexibility in data preprocessing.
- **Improved Versatility**: Enhances the model's ability to generalize across different types of training data, leading to potentially better performance and robustness in diverse applications.

### CODE REVIEW

**Comment by edkazcarlson** [comment]
I have read the CLA Document and I sign the CLA

**Comment by edkazcarlson** [comment]
Unsure if this will help with reviewing, but I tested the changes with the following code
```python
def preprocess32(x):
    """
    Convert to float32 and normalize to 0-1
    Returns: np array
    """
    if (type(x) == type(torch.tensor([]))):
        x = x.transpose(0,1).transpose(1,2) # go from channel, height, width to height, width, channel
        x = x.numpy().astype(np.float32)
        if np.max(x) > 1.1:
            x /= 255
    elif type(x) == type(Image.Image()):
        x = np.float32(np.asarray(x))
        x /= 255
    elif type(x) == type(np.array([])):
        x = np.float32(x)
        x /= 255
    else: 
        print(f'In preprocess32 found {type(x)} wanted {type(Image.Image())} or {type(np.array())} or {type(torch.tensor([]))}')
        exit()
        
    assert np.max(x) <= 1.01, f'np.max(x) {np.max(x)}'
    assert np.min(x) >= -.01, f'np.min(x) {np.min(x)}'
    x = np.clip(x, 0, 1)
    return x

def FourChannelTransformMethod(x):
    x = preprocess32(x)
    x = torch.tensor(x, dtype= torch.float)
    x = x.transpose(2,1).transpose(1,0) # hwc -> chw
    x = torch.cat((x, torch.zeros_like(x[0]).unsqueeze(0)), dim=0)
    return x #c h w 

class FourChannelTransform(object):
    """Changes an image from bgr to lrgb.

    Args: normalizeSB: boolean that is true if the saturation and brightness are normalized around 0
    """
    def __init__(self):
        pass
    def __call__(self, labels):
        img = FourChannelTransformMethod(labels['img'])
        labels['img'] = img.to(torch.float16)
        return labels
    
    
class ThreeChannelTransform(object):
    """Changes an image from bgr to lrgb.

    Args: normalizeSB: boolean that is true if the saturation and brightness are normalized around 0
    """
    def __init__(self, dtype):
        self.dtype = dtype
        pass
    def __call__(self, labels):
        labels['img'] = labels['img'].to(self.dtype)
        return labels

class TwoChannelTransform(object):
    def __init__(self):
        pass

    def __call__(self, labels):
        labels['img'] = labels['img'][0:2]
        return labels

def firstTest():
    print('Default DetectionTrainer')
    overrides = {'epochs': 2, 'imgsz': 640, 'data': 'coco.yaml', 'model': f'yolov8n.yaml', 'inputCh': 3, 'batch': 8, 'close_mosaic': 1}
    trainer = DetectionTrainer(overrides=overrides)
    trainer.train()

def secondTest():
    print('3 channel detection trainer with float16')
    overrides = {'epochs': 2, 'imgsz': 640, 'data': 'coco.yaml', 'model': f'yolov8n.yaml', 'inputCh': 3, 'batch': 8, 'close_mosaic': 1}
    trainer = DetectionTrainer(overrides=overrides, append_label_transforms=ThreeChannelTransform(torch.float16))
    trainer.train()

def thirdTest():
    print('3 channel detection trainer with float32')
    overrides = {'epochs': 2, 'imgsz': 640, 'data': 'coco.yaml', 'model': f'yolov8n.yaml', 'inputCh': 3, 'batch': 8, 'close_mosaic': 1}
    trainer = DetectionTrainer(overrides=overrides, append_label_transforms=ThreeChannelTransform(torch.float32))
    trainer.train()
    
def fourthTest():
    print('4 channel detection')
    overrides = {'epochs': 2, 'imgsz': 640, 'data': 'coco.yaml', 'model': f'yolov8n.yaml', 'inputCh': 4, 'batch': 8, 'close_mosaic': 1}
    trainer = DetectionTrainer(overrides=overrides, append_label_transforms=FourChannelTransform())
    trainer.train()
    
def fifthTest():
    print('2 channel detection')
    overrides = {'epochs': 2, 'imgsz': 640, 'data': 'coco.yaml', 'model': f'yolov8n.yaml', 'inputCh': 2, 'batch': 8, 'close_mosaic': 1}
    trainer = DetectionTrainer(overrides=overrides, append_label_transforms=TwoChannelTransform())
    trainer.train()
    
```

**Comment by glenn-jocher** [comment]
Thanks for sharing your testing code! It looks comprehensive and covers a variety of scenarios with different channel configurations and data types. This will definitely help in understanding how the changes perform across different setups. If you encounter any issues or have further suggestions, feel free to share! 🚀

**Comment by Burhan-Q** [comment]
@edkazcarlson hey, saw your message in the Discord server. This is an interesting PR to be sure, but I wouldn't be the best person to evaluate it. I did have a couple of questions for you to consider:

- With the `inputCh` argument, have you tested all models and modes with additional channels? I think this would warrant a new `pytest`. Be sure to test segmentation, pose, and OBB models as well.
    
    - Also, I would recommend to change it to `input_ch` as camelcase isn't the standard for us
    
- I see your tests above, but how would someone utilize these changes if they're using the `model.train()` method? Not necessarily something you have to reply to me with an answer for, but just wanted to bring it up. A majority of users are going to start training in this way.

- I would recommend adding some updates to the Docs. Minimum would be to update the training arguments, but would be good to see an example of how to use it with the normal `train()` method.

These are all things for you to consider, but I don't need you or expect you to reply. My suggestions are only that, suggestions. I shared them as I believe they are important for the PR to be accepted, but doesn't ensure it will or not. Please be patient, we have so much going on and only a handful of people working on PR reviews and a single person to approve/merge changes. Thanks for the PR 🚀

**Comment by edkazcarlson** [comment]
Hi @Burhan-Q Thanks for the feedback.
I had initially only changed the Classification/Detection trainer/models since those were the only ones that inherited from BaseTrainer, I can update the other models/trainers too.
Thanks for the feedback on model.train(), I usually go through the xTrainer classes so that's good to know, I'll update that too.

I'll update the docs in the same way that they were updated for the bgr config change here https://github.com/ultralytics/ultralytics/commit/33607ca2784c69c55b8d343c885976a279e47a6d#diff-9ff4a2991a2a62c7a9a42d545abea3b8626c77ae6c08cb295e20e8ee3128c3dd. For testing I'll just set up a single test with a 1 channel image, I don't want to blow up your github pipelines budget :P 

Thanks for the feedback, and I understand that you have a lot of PRs to go through, was just a bit confused because of the glennbot
